### PR TITLE
Jit: Increase FARCODE_SIZE_MMU & CODE_SIZE

### PR DESCRIPTION
**Pull request by dreamsyntax**
Increases FARCODE_SIZE_MMU from * 48 -> * 128
Increases CODE_SIZE from * 32 -> * 128

The numbers were arbitrarily chosen as I'm not sure what would be considered a 'safe' upgrade amount and how many games / system requirements this would affect. I want to get the discussion going on how this should be approached.

### Background
Note: TLDR at next section

https://forums.dolphin-emu.org/Thread-regression-vulkan-compile-shaders-before-starting-broken

Shadow the Hedgehog (GUPE8P, and any other region variant) of the game has the issue of a hard stutter occasionally while playing, regardless of any Graphics / Config setting - every possible setting was tested and ruled out, including per-game settings such as FPRF.

In most machines, the stutter will be experienced at least once within a 30 min period. It did not matter which video backend, linux/mac/windows host, nor which GPU manufacturer.

Originally, the speedrunning community of Shadow The Hedgehog thought the issue was shader or shader compilation related, but it was odd that this was the only game that seemed to be affected. No other games seem to exhibit the issue.

Fast forward to today, we investigated and realized this log was being thrown whenever the issue occurred.
```
17:05:429 Core\PowerPC\Jit64\Jit.cpp:1163 W[PowerPC]: JIT ran out of space in far code region during code generation.
17:05:429 Core\PowerPC\Jit64\Jit.cpp:808 W[PowerPC]: flushing code caches, please report if this happens a lot
```

In our speedrun approved configuration, MMU is turned on, so for testing we adjusted `FARCODE_SIZE_MMU` only.

Unfortunately, this just led to the below occurring instead:
```
20:02:610 Core\PowerPC\Jit64\Jit.cpp:1161 W[PowerPC]: JIT ran out of space in near code region during code generation.
20:02:610 Core\PowerPC\Jit64\Jit.cpp:808 W[PowerPC]: flushing code caches, please report if this happens a lot
```

After adjusting `CODE_SIZE` as well, we were able to have 3 different machines successful run the game without any stutter and none of the above logs.

Worth noting, the issue described occurs regardless if you have MMU on or not, but for testing this branch of dolphin we only verified with MMU on.

### TLDR
* Shadow the Hedgehog (GUPE8P) GameCube title has a reproducible stutter every ~30mins, due to
```JIT ran out of space in far code region during code generation.```
* Occurs with MMU ON and OFF.
* Reproducible on any host / hardware
* We do not know what the impacts of this change would be on other games, nor what 'good' values should be.

### CODE REVIEW

**Comment by JMC47** [comment]
@JosJuice What do you think of this?  I'm actually for increasing it, but I do realize it means cache clearing will take longer when it does need to be cleared.  I'm of the opinion that fewer, longer stutters is better than more shorter stutters.  And in some games, this change means we no longer end up maximizing the cache.

**Comment by JosJuice** [comment]
I suppose I'm not opposed. But we should probably up the non-MMU cache size too.

**Comment by dreamsyntax** [comment]
> @JosJuice What do you think of this? I'm actually for increasing it, but I do realize it means cache clearing will take longer when it does need to be cleared. I'm of the opinion that fewer, longer stutters is better than more shorter stutters. And in some games, this change means we no longer end up maximizing the cache.



> I suppose I'm not opposed. But we should probably up the non-MMU cache size too.

Yes, it will be adjusted for non-MMU as well. I just did not change it yet since I'm just not sure on what a good 'size' will be.

With the MMU tests we were seeing ~500mb - 1gb of RAM usage (variable).

**Comment by JosJuice** [comment]
Important note about JitArm64: The emitted code won't work correctly if the combined size of nearcode and farcode is bigger than 128 MiB, due to limited encoding space in the B instruction.

**Comment by dreamsyntax** [comment]
> Important note about JitArm64: The emitted code won't work correctly if the combined size of nearcode and farcode is bigger than 128 MiB, due to limited encoding space in the B instruction.

In that case, do you think it would be better off to not touch JitArm64 at all? Considering the size was already bigger, and I've yet to test on arm.

**Comment by dreamsyntax** [comment]
Experimenting with JitArm settings, unfortunately it IS possible to get the too large distance error at runtime / not at game boot
4eb6bff7-7b51-450a-8829-7169071244fa

**Comment by dreamsyntax** [comment]
Changed ARM code_size back to *32
Kept ARM farcode to *88;

Still can result in
```
Core/PowerPC/JitArm64/Jit.cpp:1247 W[PowerPC]: JIT ran out of space in far code region during code generation.
Core/PowerPC/JitArm64/Jit.cpp:927 W[PowerPC]: flushing code caches, please report if this happens a lot
```

But occurs ~4m 30s into runtime opposed to 1m 20s.

**Comment by dreamsyntax** [comment]
Did play test on a game with lots of known rel switching on this build for 1hr 30min, no cache clears and no stutters.

**Comment by dreamsyntax** [comment]
Removed JitArm changes.

**Comment by dreamsyntax** [comment]
@JosJuice @JMC47 testing with potentially problematic titles has shown no degradation.
Shadow game has been tested in 3hr+ sessions at this point with no stutter using this PR.
At this point I think this is ready.

**Comment by fallaha56** [comment]
Sorry if it's obvious / wrong place but quick question as I think I noticed improvements elsewhere (and I do reasonably regular updates) -is there any reason this would improve Rogue Squadron III as I now seem to be hitting 60fps vs a much more stutter filled experience on a 5950x...

**Comment by JosJuice** [comment]
Yes, this change makes JIT cache stuttering less frequent in all games that are affected by JIT cache stuttering.

**Comment by dreamsyntax** [comment]
> Sorry if it's obvious / wrong place but quick question as I think I noticed improvements elsewhere (and I do reasonably regular updates) -is there any reason this would improve Rogue Squadron III as I now seem to be hitting 60fps vs a much more stutter filled experience on a 5950x...

Just to be sure I'm not misunderstanding, is the performance WORSE with this build or better?
Rogue Squadron III has dynamic load/unloading code so there is the possibility it would get worse. I had someone with the game test it and they did not see the issue in logs.

**Comment by fallaha56** [comment]
> > Sorry if it's obvious / wrong place but quick question as I think I noticed improvements elsewhere (and I do reasonably regular updates) -is there any reason this would improve Rogue Squadron III as I now seem to be hitting 60fps vs a much more stutter filled experience on a 5950x...
> 
> Just to be sure I'm not misunderstanding, is the performance WORSE with this build or better? Rogue Squadron III has dynamic load/unloading code so there is the possibility it would get worse. I had someone with the game test it and they did not see the issue in logs.

hi performance seems *better*, i was always on the cusp of 60fps but with frame drops into mid-50s, that seems gone

thus it's my go-to game to test new versions for speed improvements etc so pretty sure

and that was already using every trick i could think off...HLE audio, VLK backend, multithreading, etc

hope this helps
